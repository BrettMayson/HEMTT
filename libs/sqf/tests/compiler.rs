#![allow(clippy::unwrap_used)]

pub use float_ord::FloatOrd as Scalar;
use hemtt_preprocessor::Processor;
use hemtt_sqf::{compiler::serializer::Compiled, parser::database::Database};
use hemtt_workspace::LayerType;

macro_rules! compile {
    ($dir: expr, $file:ident) => {
        paste::paste! {
            #[test]
            fn [<simple_ $file>]() {
                let (bin, compiled) = optimize($dir, stringify!($file));
                // bin to hex
                let hex = bin.iter().map(|b| format!("{:02x}", b)).collect::<String>();
                insta::assert_snapshot!(hex);
                insta::assert_debug_snapshot!(compiled);
            }
        }
    };
}

compile!("optimizer", consume_array);
compile!("optimizer", static_math);
compile!("optimizer", scalar);
compile!("optimizer", string_case);
compile!("simple", dev);
compile!("simple", eventhandler);
compile!("simple", foreach);
compile!("simple", format_font);
compile!("simple", get_visibility);
compile!("simple", hash_select);
compile!("simple", hello);
compile!("simple", include);
compile!("simple", oneline);
compile!("simple", semicolons);

const ROOT: &str = "tests/";

fn optimize(folder: &str, file: &str) -> (Vec<u8>, Compiled) {
    let folder = std::path::PathBuf::from(ROOT).join(folder);
    let workspace = hemtt_workspace::Workspace::builder()
        .physical(&folder, LayerType::Source)
        .finish(None, false, &hemtt_common::config::PDriveOption::Disallow)
        .unwrap();
    let source = workspace.join(format!("{file}.sqf")).unwrap();
    let processed = Processor::run(&source).unwrap();
    let mut writer = Vec::new();
    let statements = hemtt_sqf::parser::run(&Database::a3(false), &processed).unwrap();
    let compiled = statements.compile(&processed).unwrap();
    statements
        .optimize()
        .compile_to_writer(&processed, &mut writer)
        .unwrap();
    (writer, compiled)
}
