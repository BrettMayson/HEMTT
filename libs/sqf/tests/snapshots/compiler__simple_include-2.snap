---
source: libs/sqf/tests/compiler.rs
expression: compiled
---
Compiled {
    entry_point: 8,
    constants_cache_compression: true,
    constants_cache: [
        Scalar(
            1.0,
        ),
        Scalar(
            2.0,
        ),
        Scalar(
            3.0,
        ),
        Scalar(
            4.0,
        ),
        String(
            "test",
        ),
        Code(
            Instructions {
                contents: [
                    EndStatement,
                    GetVariable(
                        1,
                        SourceInfo {
                            offset: 158,
                            file_index: 1,
                            file_line: 6,
                        },
                    ),
                    Push(
                        4,
                    ),
                    CallBinary(
                        2,
                        SourceInfo {
                            offset: 161,
                            file_index: 1,
                            file_line: 6,
                        },
                    ),
                    AssignToLocal(
                        3,
                        SourceInfo {
                            offset: 140,
                            file_index: 1,
                            file_line: 6,
                        },
                    ),
                ],
                source_pointer: Source {
                    offset: 140,
                    length: 30,
                },
            },
        ),
        Scalar(
            0.0,
        ),
        String(
            "#line 1 \"\\something.hpp\"\nprivate _things = [1,2,3,4];\n\nprivate _things = [1,2,3,4];\n\nprivate _things = [1,2,3,4];\n#line 5 \"\\include.sqf\"\n{\n\tprivate thinghi = _x + \"test\";\n} forEach [0,1,2,3];\nsystemChat str _things;\n",
        ),
        Code(
            Instructions {
                contents: [
                    EndStatement,
                    Push(
                        0,
                    ),
                    Push(
                        1,
                    ),
                    Push(
                        2,
                    ),
                    Push(
                        3,
                    ),
                    MakeArray(
                        4,
                        SourceInfo {
                            offset: 0,
                            file_index: 0,
                            file_line: 0,
                        },
                    ),
                    AssignToLocal(
                        0,
                        SourceInfo {
                            offset: 25,
                            file_index: 0,
                            file_line: 1,
                        },
                    ),
                    EndStatement,
                    Push(
                        0,
                    ),
                    Push(
                        1,
                    ),
                    Push(
                        2,
                    ),
                    Push(
                        3,
                    ),
                    MakeArray(
                        4,
                        SourceInfo {
                            offset: 0,
                            file_index: 0,
                            file_line: 0,
                        },
                    ),
                    AssignToLocal(
                        0,
                        SourceInfo {
                            offset: 55,
                            file_index: 0,
                            file_line: 3,
                        },
                    ),
                    EndStatement,
                    Push(
                        0,
                    ),
                    Push(
                        1,
                    ),
                    Push(
                        2,
                    ),
                    Push(
                        3,
                    ),
                    MakeArray(
                        4,
                        SourceInfo {
                            offset: 0,
                            file_index: 0,
                            file_line: 0,
                        },
                    ),
                    AssignToLocal(
                        0,
                        SourceInfo {
                            offset: 85,
                            file_index: 0,
                            file_line: 5,
                        },
                    ),
                    EndStatement,
                    Push(
                        5,
                    ),
                    Push(
                        6,
                    ),
                    Push(
                        0,
                    ),
                    Push(
                        1,
                    ),
                    Push(
                        2,
                    ),
                    MakeArray(
                        4,
                        SourceInfo {
                            offset: 0,
                            file_index: 0,
                            file_line: 0,
                        },
                    ),
                    CallBinary(
                        4,
                        SourceInfo {
                            offset: 173,
                            file_index: 1,
                            file_line: 7,
                        },
                    ),
                    EndStatement,
                    GetVariable(
                        0,
                        SourceInfo {
                            offset: 207,
                            file_index: 1,
                            file_line: 8,
                        },
                    ),
                    CallUnary(
                        5,
                        SourceInfo {
                            offset: 203,
                            file_index: 1,
                            file_line: 8,
                        },
                    ),
                    CallUnary(
                        6,
                        SourceInfo {
                            offset: 192,
                            file_index: 1,
                            file_line: 8,
                        },
                    ),
                ],
                source_pointer: Constant(
                    7,
                ),
            },
        ),
    ],
    names_cache: [
        "_things",
        "_x",
        "+",
        "thinghi",
        "foreach",
        "str",
        "systemchat",
    ],
    file_names: [
        "/something.hpp",
        "/include.sqf",
    ],
}
